api4/post.go

```diff
@@ -316,6 +338,54 @@ func searchPosts(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(c.App.PostListWithProxyAddedToImageURLs(posts).ToJson()))
 }
 
+func hasTeamOrSystemContextPermission(teamID string, userID string, permission *model.Permission, app *app.App) bool {
+	return (app.HasPermissionToTeam(userID, teamID, permission) || app.HasPermissionTo(userID, permission))
+}
+
+func searchParamsAdjustedForPermissions(searchParams []*model.SearchParams, teamID string, channelsWithReadPermission *model.ChannelList) ([]*model.SearchParams, *model.AppError) {
+	// Collect the channel names for the given team.
+	var channelNames []string
+	for _, channel := range *channelsWithReadPermission {
+		if channel.TeamId == teamID {
+			channelNames = append(channelNames, channel.Name)
+		}
+	}
+
+	// Get the intersect of:
+	// * channals user has permission to read, and
+	// * channels user scoped to using 'in:channel-name' search parameters.
+	var intersectChannelNames []string
+	for _, searchParam := range searchParams {
+		if len(searchParam.InChannels) > 0 {
+			intersectChannelNames = intersect(searchParam.InChannels, channelNames)
+		} else {
+			intersectChannelNames = channelNames
+		}
+		searchParam.InChannels = intersectChannelNames
+	}
+
+	// If intersect of channels is empty clear the search params
+	// because there are no channels user can read (for give search).
+	if len(intersectChannelNames) == 0 {
+		searchParams = model.ParseSearchParams("")
```

Comment:
The user might not have any read_channel permissions for channels on the team but I think we should still search DMs and GMs, right? So we could probably just strip the `InChannel` bit out and return the rest of the search params instead of blanking it

Reply:
I'm having trouble understanding this comment. Could you rephrase?

Reply:
If I'm correct, `intersectChannelNames` will include all the names of channels on the team `teamID` that the user has the for `READ_CHANNEL` permission, right? So if `intersectChannelNames` is empty you're clearing the search parameters. That then gets passed down all the way to the posts store `Search()` function which will return an empty post list since search params is empty.

So that works well for the case of searching public and private channels, but this endpoint would also return results for direct/group message channels as well. DMs/GMs don't have team IDs associated with them though (it's just a blank string), so unless I'm reading the code wrong they aren't getting handled here. E.g. in a case where the user has no `READ_CHANNEL` permissions on any channel in a team, they /should/ still get search results returned for DMs/GMs but with this bit here that won't happen.

That make sense?

It think it might simply be removing this check, since `InChannel` should already be blank. You might also be able to check for blank team IDs and include those in a separate array, maybe `dmChannels`, and also check here if that array is empty

Reply:
Yes you have read the code perfectly. 

However, `read_channel` is required by all channel types (open, private, direct, and group).

If for some reason the `ChannelMember` doesn't have a role that includes `read_channel` then it stands to reason that the results should not be returned, as per [ADR 1](https://gist.github.com/mkraft/17f62c66802bf2d68806586ecede97bd).

Reply:
Right you should have `READ_CHANNEL` for every channel you can read but this will be skipping DMs/GMs because of them having a blank team ID, not because they're missing `READ_CHANNEL` (which they shouldn't be).

Reply:
@jwilander You're right. I didn't realize DM's didn't have a team_id. Good catch. I'll add a test for that and make the appropriate change for the new implementation.